blog/2020/09/rr-memory-magic.md

```diff
@@ -0,0 +1,503 @@
+Array{Any, (2,)}[
+  Array{Float64, (1,)}[13.5571],
+  0.01415]
+$3 = void
+(rr) p jl_(((jl_value_t**)0x42f54590)[20000])
+Array{Any, (2,)}[
+  Array{Float64, (1,)}[58.965],
+  1.40691]
+$4 = void
+```
+
+Alright, so we're currently trying to allocate a tuple of `42600` elements, all of which are 2-element arrays (of type `Array{Any, 1}`). Since 10000 is less than `42600`, it's also quite likely that this is the operation that we crashed in during the recording (also remember that the crash location was `jl_apply_tuple`, which is part of the allocation of a new tuple type). However, we haven't quite explained things yet. If we set a breakpoint at the location of the crash:
+
+```
+(rr) b *0x7f049e2ba8e3
+Breakpoint 1 at 0x7f049e2ba8e3: file /buildworker/worker/package_linux64/build/src/julia.h, line 1222.
+(rr) c
+Continuing.
+[FATAL /home/keno/rr/src/ReplaySession.cc:636:check_pending_sig()] 
+```
+
+we just get the divergence again, so it never actually reached this execution point, but diverged earlier (but only happened to crash there). Can we find out what happened? Looking back at the assembly, we see that r10 was loaded from
+`rax`, which we also have: rax:0x436487c0. What's there currently (remember we're still 10000 ticks before the crash):
+
+```
+(rr) p *(void**)0x436487c0
+$5 = (void *) 0x7f0490b784d0
+(rr) p jl_($5)
+Array{Any, 1}
+$6 = void
+```
+
+That's exactly the pointer we saw during the crash, but the high bit isn't set. The next thing we might want to ask is what (recorded) changes *do* happen to this memory location:
+
+```
+(rr) awatch *(void**)0x436487c0
+Hardware access (read/write) watchpoint 1: *(void**)0x436487c0
+
+(rr) c
+Continuing.
+
+Thread 1 hit Hardware access (read/write) watchpoint 1: *(void**)0x436487c0
+
+Value = (void *) 0x7f0490b784d0 <jl_system_image_data+16912>
+0x00007f049e2b584e in typekey_hash (nofail=0, n=<optimized out>, key=0x435f2dc8, 
+    tn=0x7f0490b99000 <jl_system_image_data+150848>) at /buildworker/worker/package_linux64/build/src/jltypes.c:1013
+1013  /buildworker/worker/package_linux64/build/src/jltypes.c: No such file or directory.
+
+typekey_eq (key=key@entry=0x435f2dc8, n=n@entry=46200, tt=<optimized out>, tt=<optimized out>)
+    at /buildworker/worker/package_linux64/build/src/jltypes.c:560
+560 in /buildworker/worker/package_linux64/build/src/jltypes.c
+(rr) 
+Continuing.
+[FATAL /home/keno/rr/src/ReplaySession.cc:636:check_pending_sig()] 
+ (task 19490 (rec:7760) at time 51911)
+ -> Assertion `false' failed to hold. Replaying `SIGNAL: SIGSEGV(det)': expecting tracee signal or trap, but instead at `write' (ticks: 144040993972)
+```
+
+We see some accesses from the type cache, and then the divergence. What about in the other direction? (Note that the divergence crashed the rr process, so we have to first get back to the same execution point - luckily since this is a recording, everything is stable, so we can just `seek-ticks` right back to where we were).
```

Comment:
```suggestion
We see some accesses from the type cache, and then the divergence. What about in the other direction? (Note that the divergence crashed the `rr` process, so we have to first get back to the same execution point - luckily since this is a recording, everything is stable, so we can just `seek-ticks` right back to where we were).
```